### PR TITLE
Fix likes/dislikes

### PR DESCRIPTION
**Pull request by SamantazFox**
Use new technique and improve fallbacks of like/dislike count.
thanks @TiA4f8R for the formula :)

### CODE REVIEW

**Comment by SamantazFox** [comment]
deployed on test.invidious.io, if needed, along a few other of my PRs.

**Comment by unixfox** [comment]
New bugs probably related to this PR:
- https://github.com/iv-org/invidious/issues/2664
- https://github.com/iv-org/invidious/issues/2663

**Comment by SamantazFox** [comment]
@unixfox I've pushed ceb1feb3502367ec41c8ae7e28a0cfbd7a2e2170, so you can update your patches to try it out.

File: src/invidious/videos.cr [hunk 1]
```diff
@@ -877,42 +877,84 @@ def extract_video_info(video_id : String, proxy_region : String? = nil, context_
...
+      .try &.["videoSecondaryInfoRenderer"]
+
+  # Likes/dislikes
+
+  toplevel_buttons = video_primary_renderer
+    .try &.dig?("videoActions", "menuRenderer", "topLevelButtons")
+
+  if toplevel_buttons
+    likes_button = toplevel_buttons.as_a
+      .find(&.["toggleButtonRenderer"]["defaultIcon"]["iconType"].as_s.== "LIKE")
```
Comment (matthewmcgarvey) [inline src/invidious/videos.cr:900]: Weird to see the use of the square brace access when dig is being used everywhere else. Any particular reason for this or could it be switched?

Same goes for the line fetching the dislikes button

File: src/invidious/videos.cr [hunk 2]
```diff
@@ -877,42 +877,84 @@ def extract_video_info(video_id : String, proxy_region : String? = nil, context_
...
+      LOGGER.trace("extract_video_info: Found \"likes\" button. Button text is \"#{likes_txt}\"")
+      LOGGER.debug("extract_video_info: Likes count is #{likes}") if likes
+    end
+
+    dislikes_button = toplevel_buttons.as_a
+      .find(&.["toggleButtonRenderer"]["defaultIcon"]["iconType"].as_s.== "DISLIKE")
+      .try &.["toggleButtonRenderer"]
+
+    if dislikes_button
+      dislikes_txt = (dislikes_button["defaultText"] || dislikes_button["toggledText"]?)
```
Comment (matthewmcgarvey) [inline src/invidious/videos.cr:917]:
```suggestion
      dislikes_txt = (dislikes_button["defaultText"]? || dislikes_button["toggledText"]?)
```

⚠️

File: src/invidious/videos.cr [hunk 3]
```diff
@@ -877,42 +877,84 @@ def extract_video_info(video_id : String, proxy_region : String? = nil, context_
...
-          .try &.["rows"]?
-            .try &.as_a
+  # Top level elements
+
+  primary_results = player_response
+    .dig?("contents", "twoColumnWatchNextResults", "results", "results", "contents")
+
+  video_primary_renderer = primary_results
+    .try &.as_a.find(&.["videoPrimaryInfoRenderer"]?)
+      .try &.["videoPrimaryInfoRenderer"]
```
Comment (matthewmcgarvey) [inline src/invidious/videos.cr:887]: Bunch of lines in here don't have the safe access `#[]?` but should they?

File: src/invidious/videos.cr [hunk 4]
```diff
@@ -877,42 +877,84 @@ def extract_video_info(video_id : String, proxy_region : String? = nil, context_
...
     title = row["metadataRowRenderer"]?.try &.["title"]?.try &.["simpleText"]?.try &.as_s
-    contents = row["metadataRowRenderer"]?
-      .try &.["contents"]?
-        .try &.as_a[0]?
+    contents = row.dig?("metadataRowRenderer", "contents", 0)
 
     if title.try &.== "Category"
-      contents = contents.try &.["runs"]?
-        .try &.as_a[0]?
+      contents = contents.try &.["runs"]?.try &.as_a[0]?
```
Comment (matthewmcgarvey) [inline src/invidious/videos.cr:957]:
```suggestion
      contents = contents.try &.dig?("runs", 0)
```
I believe that's the same and fits with the theme of the PR

File: src/invidious/videos.cr (hunk 1 above)
Comment (SamantazFox) [reply src/invidious/videos.cr:900]: yeah, you're right, that should use `.dig()`!

File: src/invidious/videos.cr (hunk 3 above)
Comment (SamantazFox) [reply src/invidious/videos.cr:887]: I'm not sure to understand what you mean. We try to cast `primary_results` to an array and `.find` a sub-element that has the `videoPrimaryInfoRenderer` key. Then we get to that element.
Though, yes, it's assumed that if the cast to `Array` is sucessful, it means that a `JSON::Any` element follows.

File: src/invidious/videos.cr (hunk 4 above)
Comment (SamantazFox) [reply src/invidious/videos.cr:957]: Oh, the goal of this PR wasn't to rewrite/clean the whole code around, but that doesn't hurt ^^

**Comment by SamantazFox** [comment]
@unixfox nothing to report during testing?

**Comment by unixfox** [comment]
> @unixfox nothing to report during testing?

No it seems like it's working fine :).

**Pull request merged**